File: Сourse I/Алгоритмы Python/Part2/семинары/pract3/task4.py
"""
4*.  Дано слово, состоящее только из строчных латинских букв. Проверьте, является ли это
слово палиндромом. Выведите YES или NO. При решении этой задачи нельзя пользоваться
циклами и нельзя использовать срезы с шагом, отличным от 1
Пример: radar YES
Yes No
"""

def r_palindrome(s):
    if len(s) == 1:
        return "YES"
    else:
        if s[:1] == s[-1:]:
            if len(s) == 2:
                return "YES"
            return r_palindrome(s[1:-1])
        return "NO" 

if __name__ == "__main__":
    print(r_palindrome("лимузинизумил"))

File: Сourse I/Алгоритмы Python/Part2/семинары/pract3/java/task4.java
public class Solution {
    public static String recursion(String s) {
        // Базовый случай
        if (s.length() == 1) {
            return "YES";
        } else {
            if (s.substring(0, 1).equals(s.substring(s.length() - 1, s.length()))) {
                // Базовый случай 
                if (s.length() == 2) {
                    return "YES";
                }
                // Шаг рекурсии / рекурсивное условие
                return recursion(s.substring(1, s.length() - 1));
            } else {
                return "NO";
            }
        }
    }
    public static void main(String[] args) {
        System.out.println(recursion("radar")); // вызов рекурсивной функции
    }
}